server/src/main/java/org/elasticsearch/snapshots/InFlightShardSnapshotStates.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.snapshots;
+
+import com.carrotsearch.hppc.cursors.ObjectObjectCursor;
+import org.elasticsearch.cluster.SnapshotsInProgress;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.repositories.IndexId;
+import org.elasticsearch.repositories.RepositoryShardId;
+import org.elasticsearch.repositories.ShardGenerations;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Holds information about currently in-flight shard level snapshot or clone operations on a per-shard level.
+ * Concretely, this means information on which shards are actively being written to in the repository currently
+ * as well as the latest written shard generation per shard in case there is a shard generation for a shard that has
+ * been cleanly written out to the repository but not yet made part of the current {@link org.elasticsearch.repositories.RepositoryData}
+ * through a snapshot finalization.
+ */
+public final class InFlightShardSnapshotStates {
```

Comment:
I was really torn about adding a new class for this one, but I figured adding this logic to `SnapshotsInProgress` or `SnapshotsService` isn't so great because those two classes are already very very complicated. This way we can at least take some of the complication out of them?

Reply:
I think it was a good idea to factor out this into its own class. The client code is easier to read.